vendor/github.com/SmartBFT-Go/consensus/internal/bft/controller.go

```diff
@@ -630,6 +623,10 @@ func (c *Controller) sync() (viewNum uint64, seq uint64, decisions uint64) {
 
 	latestSequence := c.latestSeq()
 
+	c.Logger.Debugf("Node %d is setting the checkpoint after sync to view %d and seq %d", c.ID, md.ViewId, md.LatestSequence)
```

Comment:
I'm not sure I agree with this [change](https://github.com/SmartBFT-Go/consensus/commit/154c86662788250a1e54b9c4ba303677399da24f). 

The unit test there uses a decision returned from a synchronization with view number 0, but the view proposal committed is the same sequence but with view 1. This should not be possible.
If view 0 committed sequence 1 then we cannot have view 1 committing this sequence again as done in the test.

When I run the test with view number 0 instead of 1, it crashes because negative view number. 

Can you perhaps fix the test to be more realistic? 

The only way to skip updating the checkpoint, is if we have advanced the view due to a view change, but the sync has returned a decision with an old view. 

Now, if we have moved to a new view, then either we have all blocks up until that new view, or we don't.

If we do, then we don't need to set the checkpoint, as we should have updated it in the past via sync or via direct consensus agreement.

If we don't, then there are some blocks that we have not committed but we did move to a new view regardless. 
However, this is not possible, because whenever we process we new view message, we check our latest decision is aligned with the latest decision in the new view message, so we will not actually move to the new view unless we have actually committed all blocks in the previous view.

Reply:
I don't understand you.

You don't agree with the decision?
or
You don't think there's a problem to be solved here at all?

Reply:
If there is a problem, the unit tests in your commit doesn't show it.

Reply:
Why? It completely replicates what I saw in person. Here's the [test.log](https://github.com/hyperledger/fabric/files/12508192/test.log) file where it shows up. Would you like me to show you the lines where it happens?

Reply:
We need to observe node 4

- 1799: got the PrePrepare
- 2045: collected 4 prepares from [1 2 7 3]
- 2103: processing a new view message < NewView
- 2249, 2254,2284: got message commit from 1 3 and 2
- 2303-2313: Starting view with number 1, sequence 1, and decisions 0
- 3873: Sync deliver view 0 last seq 0
- 3897: Deliver view 0 last seq 1 prevSeq 0
- 3929: Synchronizer returned with view number 0 but the controller is at view number 1
- 4067: Delivering to app from deliverDecision the last decision proposal
- 4068: Deliver view 0 last seq 1 prevSeq 1
- 4069: Committed sequence 1 twice

Reply:
I can tell you how to replicate it so you'll see that it's possible

Reply:
Can you make a test that replicates this scenario you described? Because you made a unit test which synchronizes a **different block** than what was committed. 

Can you make a test in [this package](https://github.com/SmartBFT-Go/consensus/blob/v2/test/basic_test.go) that reproduces this? 

I am asking because you are claiming a scenario happens which involves view change and synchronization, yet you made a unit test for the controller.

Reply:
Can you point me to the test of TestViewChangeAfterTryingToFork3053337697 ?

Reply:
[TestViewChangeAfterTryingToFork](https://github.com/SmartBFT-Go/consensus/blob/v2/test/basic_test.go#L2463)

I was unable to replicate this case on the integration test so that it reproduced in 100%.

But here are the steps in which it definitely shows up. Check out a version of this [commit](https://github.com/SmartBFT-Go/consensus/commit/071648f76967ae272f4770ae8281c17a1832db9a) and execute this command:
```bash
for i in {1..200}; do go test -count 1 -race ./...  >> test.log 2>&1; done; 
```

Оn my computer, out of 200 cycles, 1 or 2 end up with exactly the same result "Committed sequence 1 twice".

Reply:
> Can you make a test that replicates this scenario you described? Because you made a unit test which synchronizes a **different block** than what was committed.
> 
> Can you make a test in [this package](https://github.com/SmartBFT-Go/consensus/blob/v2/test/basic_test.go) that reproduces this?
> 
> I am asking because you are claiming a scenario happens which involves view change and synchronization, yet you made a unit test for the controller.

The way I determined the error was this. During synchronization, a block was received, but it did not lock to the checkpoint. Then the next delivery results in a Committed sequence twice.

After the above fix, there has never been a "Committed sequence twice" error

Reply:
I can crack the consensus
I know how to fix consensus.
I can't replicate the breakdown in a controlled manner.

Reply:
@HagarMeir could you please take a look at this?

Reply:
I understand you see a problem.
We need to be able to reproduce it.
The unit test in the controller you provided does not reproduce it, but creates a scenario that is not possible. 
I don't agree with your fix, if the sync returned with a lower view number then why do we set the checkpoint anyway? 

If there is a problem we need to be able to reproduce it and provide a better fix.

Reply:
> I understand you see a problem. We need to be able to reproduce it. The unit test in the controller you provided does not reproduce it, but creates a scenario that is not possible. I don't agree with your fix, if the sync returned with a lower view number then why do we set the checkpoint anyway?
> 
> If there is a problem we need to be able to reproduce it and provide a better fix.

1. I cannot reproduce this error in a controlled way. I have tried. On the integration test it appears 1 time out of 200.
2. Why do you say that the scenario is not possible? There is proof that it is possible. Above is the [log file](https://github.com/hyperledger/fabric/pull/4422#discussion_r1314336154). The unit test repeats exactly what is happening there, or rather the problem I saw. 
- The node fails to participate in the consensus.
- Increases the view.
- during synchronization we get a block that we don't have yet, but it has a smaller view number.
- did not set a checkpoint
3. Synchronization gave me back a block I don't have. Even if it's a block from a smaller view. But I don't have it. I set the checkpoint not the view, but the block numbers. 

I don't understand what's wrong.

Reply:
> 2. As Yacov said: If view 0 committed sequence 1 then we cannot have view 1 committing this sequence again as done in the test.

The block was received on the previous view at the same time as the view was changed. And it turns out that we don't have it. It came to us during synchronization (this process does not depend on smartbft). [Here](https://github.com/hyperledger/fabric/pull/4422#discussion_r1314341969) is a brief explanation of what is going on there.

> 3. Is it possible maybe that the sync returned something old? And we have already advanced? If it is possible then we shouldn't update the checkpoint at this stage.

No, it can't. Otherwise we would see a corresponding entry in the [log file](https://github.com/hyperledger/fabric/pull/4422#discussion_r1314336154)

Reply:
@yacovm @HagarMeir 
I think I know what you don't like. Yes, in the test, the second delivery of a block from the previous view is done from the viewChanger component of the deliverDecision function. Which in turn calls the Deliver component of MutuallyExclusiveDeliver.  In the unit test it was important for me to deliver a block from the previous view through the Deliver function of the MutuallyExclusiveDeliver component. I did it in the way available to me. 

It seems to me that it doesn't matter how the block from the previous view reaches the Deliver function of the MutuallyExclusiveDeliver component. The main thing is that it can do it, the proof of that is the log file with this case.

Reply:
@pfi79 why don't you make this PR upgrade to the SmartBFT commit before the controversial commit, and also take a look at [this fix](https://github.com/fabric-research/consensus/pull/11)?

Reply:
Let me check your decision. And if the error is corrected, I will easily and without hesitation replace my solution with yours.

Reply:
no no, the fix i am referencing has no connection to what we're discussing, it's something else.

Reply:
So your solution doesn't fix the twice-delivery error?

Reply:
Okay. I will accept your change now.
But there is one thing that is bothering me: I know how to crash a smartbft orderer with unknown consequences (tested on the bench). I really want to deprive myself of this knowledge.
Why not accept my solution for now, until a better one comes along?

Reply:
BTW I am now trying to look at the log you attached with the failure and hopefully write a test that will reach the committed sequence twice problem

Reply:
Were you able to replicate the case with a large number of test runs?

Reply:
No...I ran it many times and it did not fail

Reply:
Okay. I'm gonna roll mine right now.

Do you need an up-to-date log file with this error? I can replicate it for you.

Reply:
> Okay. I will accept your change now. But there is one thing that is bothering me: I know how to crash a smartbft orderer with unknown consequences (tested on the bench). I really want to deprive myself of this knowledge. 

How did you reproduce it in a real Fabric environment? What did you do?


> Why not accept my solution for now, until a better one comes along?

Because in order to accept a change we need to understand how it fixes a problem, and for that we need to know exactly how the problem manifests and why.

Reply:
Did you run it that way?

```bash
for i in {1..200}; do go test -count 1 -race ./...  >> test.log 2>&1; done; 
```

When I just ran it without that command, it failed to replicate. Everything completed successfully.

Reply:
> How did you reproduce it in a real Fabric environment? What did you do?

During actual operation, terminated and after some time started 3 out of 7 nodes. Never touched the leader. That is, I emulated a failure, such that at some point the consensus was lost, but then recovered.

Reply:
> How did you reproduce it in a real Fabric environment? What did you do?

This [test](https://github.com/SmartBFT-Go/consensus/blob/v2/test/basic_test.go#L2463) is a representation of what we had going on in the booth. And how it led to bad situations.

- first node 6 removed itself from the consensus.
- then terminated 3 nodes including (4, 5, 6)
- sent the transaction
- wait for a message from nodes (0, 1, 2, 3) that they want to change the leader.
- include (4, 5, 6) nodes
- wait for the actual change of leader and view
- wait for the transaction delivery

But there are some interesting effects here, one of which is " twice delivery".

Most likely it turns out that the leader change and the transaction solution occur simultaneously

Reply:
@yacovm done

Reply:
Pipeline for raft is hung up. Can I ask you to restart pipelines for raft?

Reply:
Thanks